### PR TITLE
Highlight Cooldowns

### PR DESCRIPTION
**Pull request by CircuitSacul**
This makes a few changes to improve highlights
 - [x] If a message contains multiple triggers, combine them into a single DM.
 - [x] Have a cooldown of 30s per (highlight, user, channel)
 - [x] If a user sends a message in a channel, don't send them any highlights for the next minute.

These are just my feel for how it should work. If you think it should be tweaked lmk.

### CODE REVIEW

**Comment by MithicSpirit** [comment]
These changes sound good to me, though please make the two cooldowns relatively easy to configure (like `MAX_HIGHLIGHTS` and `MAX_HIGHLIGHT_LENGTH`).

Might be interesting for us to look into having a proper config file for these things in the future (maybe we can commandeer the `config.json` for these purposes?).

**Comment by CircuitSacul** [comment]
> These changes sound good to me, though please make the two cooldowns relatively easy to configure (like `MAX_HIGHLIGHTS` and `MAX_HIGHLIGHT_LENGTH`).
> 
> Might be interesting for us to look into having a proper config file for these things in the future (maybe we can commandeer the `config.json` for these purposes?).

Well these things would undoubtedly be defined in config.json (with defaults in config.py). Unless you meant per-user configuration? I think that's possible too.

**Comment by Lunarmagpie** [comment]
> Well these things would undoubtedly be defined in config.json (with defaults in config.py). Unless you meant per-user configuration? I think that's possible too.

I think moving all config options to `config.py` would be a good idea. May open a PR for that before this one gets merged.

**Comment by MithicSpirit** [comment]
> Well these things would undoubtedly be defined in config.json (with defaults in config.py). Unless you meant per-user configuration? I think that's possible too.

`MAX_HIGHLIGHTS` and `MAX_HIGHLIGHT_LENGTH` aren't in `config.json` :p. As for per-user configuration, it would be nice but probably overkill for now.

**Comment by CircuitSacul** [comment]
> `MAX_HIGHLIGHTS` and `MAX_HIGHLIGHT_LENGTH` aren't in `config.json` :p. As for per-user configuration, it would be nice but probably overkill for now.

Oh oof. They should've been. Well, I'll be putting these settings in config. And yeah, per-user config is outside the scope of this pr anyways

**Comment by CircuitSacul** [comment]
This is done. Ready for review.

File: mcodingbot/plugins/highlights.py [hunk 1]
```diff
@@ -166,18 +173,49 @@ async def _dm_user_highlight(
...
-                        user_id=user_id,
-                    )
-                )
+                highlights[user_id].append(highlight)
+
+    for user_id, hls in highlights.items():
+        if (
+            sent_message_cooldown.get_retry_after((user_id, event.channel_id))
+            != 0
+        ):
```
Comment (Lunarmagpie) [inline mcodingbot/plugins/highlights.py:213]: I think the `!= 0` can be removed. It might fit on one line then.

File: mcodingbot/config.py [hunk 2]
```diff
@@ -24,6 +24,10 @@ class Config:
     patron_role: int = -1
     donor_role: int = -1
 
+    # capacity, cooldown
+    message_sent_cooldown: tuple[float, float] = (1, 60)
```
Comment (Lunarmagpie) [inline mcodingbot/config.py:28]: I think it is confusing what the `message_sent_cooldown` is for
```suggestion
    highlight_message_sent_cooldown: tuple[float, float] = (1, 60)
```

**Review by Lunarmagpie — CHANGES_REQUESTED** [review]
I think switching to `NamedTuple` for buckets would help reduce mistakes when editing this code in the future.

**Comment by Lunarmagpie** [comment]
Still would prefer if `NamedTuple` was used.

**Review by Lunarmagpie — APPROVED** [review]

**Review by Lunarmagpie — CHANGES_REQUESTED** [review]
Actually the bucket should probably be `SentMessageBucket` so its obvious what variable its used with

**Comment by CircuitSacul** [comment]
> Actually the bucket should probably be `SentMessageBucket`

I just named it "UserChannelBucket" b/c that's what it is literally

**Comment by Lunarmagpie** [comment]
> I just named it "UserChannelBucket" b/c that's what it is literally

No reason to explain the properties of the class. Makes more sense to name it so its clear where its being used.

**Review by Lunarmagpie — APPROVED** [review]

**Review by trag1c — APPROVED** [review]

**Pull request merged**